### PR TITLE
[artifactory] Add NetworkPolicy support

### LINKED ISSUE #69: [Feature] Add NetworkPolicy to each chart to restrict traffic between pods
**Is this a request for help?**:

No

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
FEATURE REQUEST:
To have better control and secure structure of traffic between each application's pods, every chart has to also deploy a NetworkPolicy descriptor with the matching rules.

- Deny all
- Allow traffic from pods in same namespace on application ports
- Allow traffic from external resources to public endpoint only

**Version of Helm and Kubernetes**:
Any

**Which chart**:
All

### PR DESCRIPTION
**Pull request by zendril**
#### PR Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Chart Version bumped
- [x] CHANGELOG.md updated
- [x] Variables and other changes are documented in the README.md

**What this PR does / why we need it**:
Many clusters are starting to employ a DenyAll default network policy and ask teams to be specific about their network policy ingress and egress.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #69 

**Special notes for your reviewer**:
Still a WIP (need to documentation and chart version bumps) but the basic template and values are working.

### CODE REVIEW

**Comment by eldada** [comment]
@zendril - this is great idea. I've had this in my backlog for a while... Yes. We do need to add this support as it's also expected to be used for hardening deployments.
- Can you give more detailed examples and document in `README.md`?
- Can you add this functionality to all charts?

**Comment by rimusz** [comment]
+1 for `Can you give more detailed examples and document in README.md?`
In this PR I think both artifactory charts should be enough.
New PR for the rest of charts :)

**Comment by zendril** [comment]
Thanks.

Will do.. Just checking to see if there was anything major I was missing before I started on the docs. 
Have it working here just fine so I think we are good. 

I'll start on docs and readme.md now.

**Comment by rimusz** [comment]
also the chart version needs a bump.

**Comment by zendril** [comment]
> also the chart version needs a bump.

yeah.. I'll take care of all the normal PR things.

**Comment by zendril** [comment]
@rimusz @eldada 
Done.

Essentially it allows multiple NetworkPolicy objects to be defined/created by specifying them as a list in values.yaml.
Defaults to 'allow all', open ingress and egress traffic (the way it is today).
Because of the complexity in the matching logic and to/from rules I decided to just let those be specified in plain yaml that matches the underlying network policy format.

File: stable/artifactory-ha/values.yaml [hunk 1]
```diff
@@ -56,6 +56,27 @@ ingress:
...
+  # Allows all ingress and egress
+  - name: artifactory
+    podSelector:
+      matchLabels:
+        app: artifactory-ha
+    egress:
+    - {}
+    ingress:
+    - {}
+  # Uncomment to allow communicate to postgresql (if postgres.enabled is true)
```
Comment (eldada) [inline stable/artifactory-ha/values.yaml]: Shouldn't it be "Uncomment to allow only artifactory pods to communicate with postgresql (if postgres.enabled is true)"

File: stable/artifactory/README.md [hunk 2]
```diff
@@ -186,6 +186,38 @@ helm install --name artifactory --set artifactory.license.secret=artifactory-lic
...
+For podSelector, ingress and egress, if nothing is provided then a default `- {}` is applied which is to allow everything.
+
+A full (but very wide open) example that results in 2 NetworkPolicy objects being created:
+```
+networkpolicy:
+  # Allows all ingress and egress to/from artifactory primary and member pods.
+  - name: artifactory
+    podSelector:
+      matchLabels:
+        app: artifactory-ha
```
Comment (eldada) [inline stable/artifactory/README.md]: `artifactory` only. No `-ha`

File: stable/artifactory/README.md [hunk 3]
```diff
@@ -186,6 +186,38 @@ helm install --name artifactory --set artifactory.license.secret=artifactory-lic
...
+  # Allows connectivity from artifactory-ha pods to postgresql pods, but no traffic leaving postgresql pod.
+  - name: postgres
+    podSelector:
+      matchLabels:
+        app: postgresql
+    ingress:
+    - from:
+      - podSelector:
+          matchLabels:
+            app: artifactory-ha
```
Comment (eldada) [inline stable/artifactory/README.md]: `artifactory` only. No `-ha`

File: stable/artifactory/README.md [hunk 4]
```diff
@@ -186,6 +186,38 @@ helm install --name artifactory --set artifactory.license.secret=artifactory-lic
...
+  # Allows all ingress and egress to/from artifactory primary and member pods.
+  - name: artifactory
+    podSelector:
+      matchLabels:
+        app: artifactory-ha
+    egress:
+    - {}
+    ingress:
+    - {}
+  # Allows connectivity from artifactory-ha pods to postgresql pods, but no traffic leaving postgresql pod.
```
Comment (eldada) [inline stable/artifactory/README.md]: `artifactory` only. No `-ha`

File: stable/artifactory/README.md [hunk 5]
```diff
@@ -186,6 +186,38 @@ helm install --name artifactory --set artifactory.license.secret=artifactory-lic
...
+NetworkPolicy specifies what ingress and egress is allowed in this namespace. It is encouraged to be more specific whenever possible to increase security of the system.
+
+In the `networkpolicy` section of values.yaml you can specify a list of NetworkPolicy objects.
+
+For podSelector, ingress and egress, if nothing is provided then a default `- {}` is applied which is to allow everything.
+
+A full (but very wide open) example that results in 2 NetworkPolicy objects being created:
+```
+networkpolicy:
+  # Allows all ingress and egress to/from artifactory primary and member pods.
```
Comment (eldada) [inline stable/artifactory/README.md]: `artifactory` only. No `primary` and `member`.

File: stable/artifactory/values.yaml [hunk 6]
```diff
@@ -57,6 +57,27 @@ ingress:
...
+  # Allows all ingress and egress
+  - name: artifactory
+    podSelector:
+      matchLabels:
+        app: artifactory
+    egress:
+    - {}
+    ingress:
+    - {}
+  # Uncomment to allow communicate to postgresql (if postgres.enabled is true)
```
Comment (eldada) [inline stable/artifactory/values.yaml]: Shouldn't it be "Uncomment to allow only artifactory pod to communicate with postgresql (if postgres.enabled is true)"

File: stable/artifactory/templates/artifactory-networkpolicy.yaml [hunk 7]
```diff
@@ -0,0 +1,34 @@
+{{- range .Values.networkpolicy }}
+apiVersion: extensions/v1beta1
+kind: NetworkPolicy
+metadata:
+  name: {{ template "artifactory-ha.fullname" $ }}-{{ .name }}-networkpolicy
```
Comment (eldada) [inline stable/artifactory/templates/artifactory-networkpolicy.yaml]: Remove all `-ha`. This is failing the CI.

**Comment by zendril** [comment]
Good catches. Thanks for being thorough. Will update and rebase in a couple of hours.

File: stable/artifactory-ha/values.yaml (hunk 1 above)
Comment (zendril) [reply stable/artifactory-ha/values.yaml]: Done

File: stable/artifactory/values.yaml (hunk 6 above)
Comment (zendril) [reply stable/artifactory/values.yaml]: Done

File: stable/artifactory/README.md (hunk 2 above)
Comment (zendril) [reply stable/artifactory/README.md]: done

File: stable/artifactory/README.md (hunk 3 above)
Comment (zendril) [reply stable/artifactory/README.md]: done

File: stable/artifactory/README.md (hunk 4 above)
Comment (zendril) [reply stable/artifactory/README.md]: Done.

File: stable/artifactory/README.md (hunk 5 above)
Comment (zendril) [reply stable/artifactory/README.md]: Done.

File: stable/artifactory/templates/artifactory-networkpolicy.yaml (hunk 7 above)
Comment (zendril) [reply stable/artifactory/templates/artifactory-networkpolicy.yaml]: done

**Comment by zendril** [comment]
@eldada Done. Review quick before the re-base monster gets me again. ;)

**Comment by zendril** [comment]
Any idea what travis is having an issue with? It is just a flaky lint?

**Comment by jainishshah17** [comment]
@zendril Your last commit broke CI. Can you please fix it.

**Comment by zendril** [comment]
> @zendril Your last commit broke CI. Can you please fix it.

Yeah.. sorry, had some trailing whitespace in the values.yaml. 
Hopefully Travis and his best buddy Lint are happy now.

**Comment by zendril** [comment]
@eldada @rimusz Looks like CI is happy again.

**Pull request merged**